### PR TITLE
[VPlan] Add scalar inferencing support for Not and Or insns

### LINKED ISSUE #87394: [LoopVectorize][VPlan] Unreachable executed "Unhandled opcode!"
Reduced LLVM IR:
```llvm ir
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "riscv64-unknown-linux-gnu"

define i32 @main() #0 {
entry:
  br label %for.body

for.cond.cleanup.loopexit:                        ; preds = %cond.end
  ret i32 0

for.body:                                         ; preds = %cond.end, %entry
  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %cond.end ]
  %conv3 = zext i1 false to i64
  %conv4 = trunc i64 %conv3 to i16
  br i1 false, label %cond.false, label %cond.end

cond.false:                                       ; preds = %for.body
  br label %cond.end

cond.end:                                         ; preds = %cond.false, %for.body
  %indvars.iv.next = add i64 %indvars.iv, 1
  %cmp = icmp ult i64 %indvars.iv, 1
  br i1 %cmp, label %for.body, label %for.cond.cleanup.loopexit
}

attributes #0 = { "target-features"="+64bit,+v" }
```

Backtrace:
```
Unhandled opcode!
UNREACHABLE executed at /scratch/tc-testing/tc-apr-2/llvm/llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:58!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt --passes=loop-vectorize reduced.ll
 #0 0x00005f05c836eb60 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x2d84b60)
 #1 0x00005f05c836bf6f llvm::sys::RunSignalHandlers() (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x2d81f6f)
 #2 0x00005f05c836c0c5 SignalHandler(int) Signals.cpp:0:0
 #3 0x000071a702442520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x000071a7024969fc __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #5 0x000071a7024969fc __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #6 0x000071a7024969fc pthread_kill ./nptl/pthread_kill.c:89:10
 #7 0x000071a702442476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #8 0x000071a7024287f3 abort ./stdlib/abort.c:81:7
 #9 0x00005f05c82a2afe (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x2cb8afe)
#10 0x00005f05c74dfe67 llvm::VPTypeAnalysis::inferScalarTypeForRecipe(llvm::VPInstruction const*) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1ef5e67)
#11 0x00005f05c74dc88c llvm::VPTypeAnalysis::inferScalarType(llvm::VPValue const*) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1ef288c)
#12 0x00005f05c750790e simplifyRecipes(llvm::VPlan&, llvm::LLVMContext&) VPlanTransforms.cpp:0:0
#13 0x00005f05c750e606 llvm::VPlanTransforms::optimize(llvm::VPlan&, llvm::ScalarEvolution&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1f24606)
#14 0x00005f05c73ded66 llvm::LoopVectorizationPlanner::buildVPlansWithVPRecipes(llvm::ElementCount, llvm::ElementCount) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1df4d66)
#15 0x00005f05c73e5319 llvm::LoopVectorizationPlanner::plan(llvm::ElementCount, unsigned int) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1dfb319)
#16 0x00005f05c73e88a9 llvm::LoopVectorizePass::processLoop(llvm::Loop*) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1dfe8a9)
#17 0x00005f05c73eb73e llvm::LoopVectorizePass::runImpl(llvm::Function&, llvm::ScalarEvolution&, llvm::LoopInfo&, llvm::TargetTransformInfo&, llvm::DominatorTree&, llvm::BlockFrequencyInfo*, llvm::TargetLibraryInfo*, llvm::DemandedBits&, llvm::AssumptionCache&, llvm::LoopAccessInfoManager&, llvm::OptimizationRemarkEmitter&, llvm::ProfileSummaryInfo*) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1e0173e)
#18 0x00005f05c73ec88d llvm::LoopVectorizePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x1e0288d)
#19 0x00005f05c62f3276 llvm::detail::PassModel<llvm::Function, llvm::LoopVectorizePass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0xd09276)
#20 0x00005f05c8195141 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x2bab141)
#21 0x00005f05c62eabd6 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0xd00bd6)
#22 0x00005f05c8193e3b llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x2ba9e3b)
#23 0x00005f05c62f2dc6 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0xd08dc6)
#24 0x00005f05c8191cb1 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x2ba7cb1)
#25 0x00005f05c5b6cce5 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x582ce5)
#26 0x00005f05c5b5f316 optMain (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x575316)
#27 0x000071a702429d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#28 0x000071a702429e40 call_init ./csu/../csu/libc-start.c:128:20
#29 0x000071a702429e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#30 0x00005f05c5b54f95 _start (/scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt+0x56af95)
zsh: IOT instruction (core dumped)  /scratch/tc-testing/tc-apr-2/build-rv64gcv/build-llvm-linux/bin/opt
```
Godbolt: https://godbolt.org/z/bvY138WfG

Triggered assertion: https://github.com/llvm/llvm-project/blob/f5991161529511ca6ebc058da2a0507c2fc5283e/llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp#L58

Found via fuzzer

### PR DESCRIPTION
**Pull request by patrick-rivos**
Fixes #87394.

### CODE REVIEW

**Comment by patrick-rivos** [comment]
VPlan so @fhahn
Please let me know if this isn't the right approach and I can dig deeper.
There are likely other missing cases but those can be added as the fuzzer finds them.

File: llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+; CHECK-NEXT:    [[CMP:%.*]] = icmp ult i64 [[INDVARS_IV]], 1
+; CHECK-NEXT:    br i1 [[CMP]], label [[FOR_BODY]], label [[FOR_COND_CLEANUP_LOOPEXIT]], !llvm.loop [[LOOP3:![0-9]+]]
+;
+entry:
+  br label %for.body
+
+for.cond.cleanup.loopexit:                        ; preds = %cond.end
+  ret i32 0
+
+for.body:                                         ; preds = %cond.end, %entry
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:42]: nit: something like `loop.header` would be slightly more descriptive

File: llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %cond.end ]
+  %zext.0 = zext i1 false to i64
+  %dead.insn = trunc i64 %zext.0 to i16
+  br i1 false, label %cond.false, label %cond.end
+
+cond.false:                                       ; preds = %for.body
+  br label %cond.end
+
+cond.end:                                         ; preds = %cond.false, %for.body
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:51]: nit: something like `loop.latch` would be slightly more descriptive

File: llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+entry:
+  br label %for.body
+
+for.cond.cleanup.loopexit:                        ; preds = %cond.end
+  ret i32 0
+
+for.body:                                         ; preds = %cond.end, %entry
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %cond.end ]
+  %zext.0 = zext i1 false to i64
+  %dead.insn = trunc i64 %zext.0 to i16
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:45]: would it be possible to expose the issue without a dead instruction to make the test overall more robust w.r.t to future changes?

File: llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+;
+entry:
+  br label %for.body
+
+for.cond.cleanup.loopexit:                        ; preds = %cond.end
+  ret i32 0
+
+for.body:                                         ; preds = %cond.end, %entry
+  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %cond.end ]
+  %zext.0 = zext i1 false to i64
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/vplan-infer-not-or-type.ll:44]: To make the test more robust w.r.t. future changes, would is it possible to reproduce the issue with a non-loop invariant/constant `zext` here?

File: llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp [hunk 5]
```diff
@@ -44,8 +45,9 @@ Type *VPTypeAnalysis::inferScalarTypeForRecipe(const VPInstruction *R) {
     CachedTypes[OtherV] = ResTy;
     return ResTy;
   }
+  case VPInstruction::Not:
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanAnalysis.cpp:48]: `Not` should only be used to negate booleans (or vectors of bools); Would probably be better to return `IntegerType::get(Ctx, 1);` like for `FCmp/ICmp` and assert that the inferred type of the operand is also `i1`.

(The reason why `PtrAdd` doesn't return `PointerType::get()` but infers the type of the pointer arg is due to address spaces)

**Review by fhahn — COMMENTED** [review]
Thanks for the patch!

**Comment by patrick-rivos** [comment]
Thanks for the review!

I split out the not and added an assert. Thankfully the fuzzer has quite a few finds of this ice so I was able to find another testcase that doesn't have any dead variables. The loop-constant seems to be a bit more challenging to get rid of.
I'll check the remaining testcases again just to make sure I didn't miss one.

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

**Comment by patrick-rivos** [comment]
Thank you! I don't have write permission to llvm - could you land this for me?

**Pull request merged**